metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java

```diff
@@ -498,6 +498,23 @@ public UsableBroker next() {
         }
     }
 
+    /**
+     * Check if the given broker is active. A broker is active if it is unfenced,
+     * and not shutting down.
```

Comment:
I'm still inclined to add another state to the broker's registration to represent the fact that it is shutting down. This seems useful for the cluster to know as a whole. At a minimum, it may be useful for debugging, but it it would also prevent unnecessary AlterPartition requests from getting sent to the controller.

Reply:
If preventing unnecessary AlterPartition requests from getting sent to the controller is the main concern, I think that we have to consider the two cases where that could happen: 1) broker controlled shutdown; and 2) broker fencing. At the moment, partition change records are written followed by the fenced record. In the controlled shutdown case, the fenced records is not written immediately but only when the broker is given the permission to shutdown.

Adding the shutting down state to the broker's registration improves 1) because it reduces the time window during which unnecessary AlterPartition requests could be sent. However it does not help for 2). For 2) we could also accept that the time window is small anyway because the fenced record is written immediately after the partition changes so it might be sufficient. Otherwise, we could perhaps write the fence records first in this case, or add another state. What's your take on 2)?

I totally agree with the debugging argument.

Reply:
Adding another state to the broker's registration will be quite easy after we merge https://github.com/apache/kafka/pull/12195.

Reply:
Yeah, good point about 2). On the other hand, perhaps this is a rarer case. The scenario requires the broker to stop heartbeating to the controller, but continue fetching. So some kind of network partition, for example. Even then, it is a race condition with receiving the FencedBroker record. Since we would still have the protection on the controller side, perhaps that is enough?

Reply:
Yeah, that seems to be a reasonable tradeoff for 2). We could also improve this later on if it turns out the be an issue. I believe that receiving the FencedBroker record will be faster for 2) than for 1) in general.

Reply:
I am a bit concerned by the controlled shutdown case because there could be a few seconds before the fenced record in written. The fenced record is written when the shutting-down broker heartbeats if all the other brokers in the cluster have caught up to the shutting down offset. I have looked into a few clusters. It typically takes 2-4 seconds before the controller grants the controlled shutdown and grants the shutdown. I have seen cases where it took 6 seconds and I suppose that it could take more time depending on the number of partitions and brokers in the cluster. That is a pretty large time window that is work optimising for in my opinion.